rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,94 @@
+# [RFC] Add collective_broadcast to the StableHLO specification
+
+Status: Review<br/>
+Initial version: 10/17/20223<br/>
+Last updated: 10/17/2023<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/1809)
+
+## Motivation
+
+StableHLO currently has [five collective communication primitives](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#collective-ops): `collective_permute`, `all_gather`, `all_to_all`, `all_reduce`, and `reduce_scatter`. However, one of the major collective communication primitives, `broadcast`, is missing from this list. This primitive allows for a one-to-many replication of a tensor to many devices efficiently. `broadcast` is a primitive in [MPI](https://www.open-mpi.org/doc/v4.1/man3/MPI_Bcast.3.php), [NCCL](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#c.ncclBroadcast), and [PyTorch](https://pytorch.org/docs/stable/distributed.html#torch.distributed.broadcast). From here on out, we will refer to this operation as `collective_broadcast` for reasons discussed later.
+
+While it technically would be possible to replicate a broadcast with a conditional mask and a `psum`, that reduces to an `all_reduce` communication primitive, which is significantly more expensive than a simple `collective_broadcast`. Additionally, when dealing with network-switch environments, the explicit use of `collective_broadcast` allows the switch to greatly optimize it's throughput when replicating to many targets simultaneously. However, XLA currently has no ability to lower directly to a mesh's `collective_broadcast` primitive, so a lot of that optimization is left on the table. 
+
+Additionally, a new compiler pass that detects usage of the old `psum` hack and replaces it with a `collective_broadcast` could be implemented only once and forever be supported by all hardware, future and current. This could have positive knock-on effects for users who don't even realize they're using it!
+
+`collective_broadcast` can be used to quickly replicate a tensor across an entire mesh, and would use less communication resources as compared to `all_gather` or `psum`. `collective_broadcast` is also the base primitive used in the [SUMMA](https://www.netlib.org/lapack/lawnspdf/lawn96.pdf) distributed GEMM algorithm. As AI computing grows larger, there likely will grow a need for these 2D distributed GEMM algorithms. Adding support for one of the needed primitives could help advance research in these areas. 
+
+## Alternatives considered
```

Comment:
I'd assume "use existing `psum` to `all_reduce` lowering and handle `all_reduce`" would be an alternative as well. PT/XLA emits a similar `all_reduce` pattern for torch bradcast. Above you mention:

> While it technically would be possible to replicate a broadcast with a conditional mask and a `psum`, that reduces to an `all_reduce` communication primitive, which is significantly more expensive than a simple `collective_broadcast`.

Could you add more details to this? Is it difficult (or infeasible) to pattern match / map the existing lowering to the desired broadcast operation on the compiler side, or why can't this operation be used for the desired optimization? It sounds like lowering this `psum` pattern to `collective_broadcast` is an option, so it would be good to add details on why we'd like a new op instead.

Reply:
>  it difficult (or infeasible) to pattern match / map the existing lowering to the desired broadcast operation on the compiler side.

The issue is that doing this pattern matching without the `collective_broadcast` primitive means that the logic will need to be reimplemented for every hardware implementation (CPU, GPU, TPU, etc). If we instead had the primitive, then we could implement that compiler pass at an early, hardware agnostic stage. Each hardware specific compiler would need to only worry about implementing `collective_broadcast`, and then they would also get the `psum hack -> broadcast` transform for free forever!

I'm aware the PT emits a similar pattern when it lowers its broadcast, and that is also one of the reasons I want to add this primitive so that PT doesn't need to do emit hack anymore.